app/controllers/bridges/oai/oai_controller.rb

```diff
@@ -0,0 +1,268 @@
+      def render_oai_error(error)
+        builder = Nokogiri::XML::Builder.new(encoding: 'UTF-8') do |xml|
+          oai_response_template(xml, include_params: %w[badVerb badArgument].exclude?(error.code)) do
+            xml.error error.message, code: error.code
+          end
+        end
+        render xml: builder.to_xml
+      end
+
+      def export_record_header(xml, task)
+        xml.header do
+          xml.identifier task.uuid
+          xml.datestamp task.updated_at.iso8601
+          xml.setSpec 'task'
+          task.labels.each do |label|
+            xml.setSpec label_to_set_spec(label)
+          end
+        end
+      end
+
+      def export_record(xml, task, format)
+        xml.record do
+          export_record_header(xml, task)
+
+          xml.metadata do
+            if format == :oai_dc
+              DublinCoreService::ExportDublinCore.call(task:, xml:)
+            elsif format == :lom
+              LomService::ExportLom.call(task:, xml:)
+            end
+          end
+        end
+      end
+
+      def find_record!(params)
+        raise OaiError.new('Missing identifier parameter', 'badArgument') unless params[:identifier]
+
+        task = Task.includes(:labels).access_level_public.find_by(uuid: params[:identifier])
+        return task if task.present?
+
+        raise OaiError.new('No record found for the specified identifier', 'idDoesNotExist')
+      end
+
+      def find_records!(params)
+        tasks = Task.includes(:labels).access_level_public.where(updated_at: parse_time_bounds!(params))
+
+        tasks = filter_tasks_by_set(tasks, params[:set]) if params[:set].present?
+        return tasks if tasks.any?
+
+        raise OaiError.new('No records matched the specified timeframe/identifier/set', 'noRecordsMatch')
+      end
+
+      def filter_tasks_by_set(tasks, set_spec)
+        raise OaiError.new("The setSpec '#{set_spec}' is invalid", 'badArgument') unless valid_set_spec?(set_spec)
+
+        return tasks if set_spec.casecmp('task').zero?
+        return tasks.where(id: TaskLabel.select(:task_id)) if set_spec.casecmp('task:label').zero?
+
+        slices = set_spec.downcase.split(':')
+        tasks.where(id: TaskLabel.select(:task_id).where(label_id: slices[2].to_i))
+      end
+
+      def valid_set_spec?(set_spec)
+        set_spec.downcase.match?(/^task$|^task:label$|^task:label:[0-9]+$/)
+      end
+
+      def oai_params
+        params.permit(:verb, :identifier, :from, :until, :metadataPrefix, :set)
+      end
+
+      def parse_time_bounds!(params)
+        tfrom = nil
+        tuntil = nil
+
+        if params[:from]
+          begin
+            tfrom = DateTime.strptime(params[:from], '%Y-%m-%d')
+          rescue Date::Error
+            begin
+              tfrom = Time.iso8601(params[:from])
+            rescue ArgumentError
+              raise OaiError.new("The specified time '#{params[:from]}' is not in iso8601 format", 'badArgument')
+            end
+          end
+        end
```

Comment:
You are right that ```DateTime.iso8601(params[:from])``` works for the :from parameter but I think for the :until parameter it does not because we only want to round to the end of the day if the :until parameter is in day granularity.

Reply:
Okay, I see. Then, I would strongly argue for persisting this reason with a comment in the code, because otherwise no one will know.

Reply:
I added one.